packages/node-app/src/controllers/giveaway.controller.ts

```diff
@@ -67,50 +85,78 @@ export const createGiveaway = async (req: Request, res: Response) => {
     res.status(500).json({ error: error.message });
   } finally {
     // remove image from tmp folder
-    fs.unlink(file.path, () => { return });
+    fs.unlink(file.path, () => {
+      return;
+    });
   }
 };
 
 export const updateGiveaway = async (req: Request, res: Response) => {
-  const file = req.file as Express.Multer.File
+  const file = req.file as Express.Multer.File;
   try {
     const { id } = req.params;
     const { title, description, prize, rules } = req.body;
-    const image = file ? fileToBase64(file as Express.Multer.File) : undefined
-    const updateFields = getDefinedFields({ title, description, prize, rules, image });
-    const giveaway = await Giveaway.findByIdAndUpdate(id, updateFields, { new: true });
+    const image = file ? fileToBase64(file as Express.Multer.File) : undefined;
+    const updateFields = getDefinedFields({
+      title,
+      description,
+      prize,
+      rules,
+      image,
+    });
+    const giveaway = await Giveaway.findByIdAndUpdate(id, updateFields, {
+      new: true,
+    });
 
     res.status(200).json(giveaway);
   } catch (error) {
     res.status(500).json({ error: error.message });
   } finally {
-    if (file) fs.unlink(file.path, () => { return });
+    if (file)
+      fs.unlink(file.path, () => {
+        return;
+      });
   }
 };
 
 export const addParticipant = async (req: Request, res: Response) => {
   try {
-    const { id } = req.params;
-    const { participant } = req.body;
-    const giveaway = await Giveaway.findById(id);
-
-    if (!giveaway)
-      return res.status(404).json({ error: 'Giveaway not found' });
-
-    if (!isValidParticipant(participant, giveaway.requirements))
-      return res.status(400).json({ error: "Participant is not whitelisted" });
-
-    const address = getParticipantAddress(participant);
-    giveaway.participants.push(address);
+    // check if giveaway exists
+    const giveaway = await Giveaway.findById(req.params.id);
+    if (!giveaway) return res.status(404).json({ error: 'Giveaway not found' });
+
+    // get participant state based on requirements and save to db
+    const participant = getParticipant(req.body);
+    const state = validateParticipant(participant, giveaway);
+    giveaway.participants.push({ id: participant.id, state });
     await giveaway.save();
```

Comment:
Maybe we should only save the new participant after the addition being successful on the contract, otherwise it will be in the local db and not on the contract

Reply:
We need to save the participant even if we don't add them the to the contract to allow for manual validation